src/utilities/modules/vaccines.js

```diff
@@ -0,0 +1,152 @@
+/**
+ * mSupply Mobile
+ * Sustainable Solutions (NZ) Ltd. 2019
+ */
+
+/**
+ * Utility and helper methods for the vaccine
+ * module.
+ */
+
+/**
+ * Extracts breaches from a set of sensor logs.
+ * Breach: Sequential sensorLog objects whose temperature is outside
+ * the lower or uppound limits, form a breach. Each array returned
+ * contains all sensorLogs which form a breach INCLUDING the sensorLog
+ * prior to the breach, and after the breach, if they exist. There may
+ * be duplicate non-breach sensorLogs as breach delimiters if there is
+ * only one sensorLog between breaches.
+ *
+ * Method sorts the sensorLogs by timestamp, iterates and pushes onto a
+ * stack. Only one unBreached sensorLog is kept at the tail of the stack.
+ * All breached sensorLogs are pushed, until a non breached log, where this
+ * marks a breach has been found. Store these sensorLogs and reset the stack.
+ *
+ * @param  {Realm.results} sensorLogs array of sensorLog objects
+ * @return {Array<Realm.results<SensorLog>} A 2D array of Realm.results. Example:
+ * [ [ sensorLog1, sensorLog2 .. ], [ sensorLog1, sensorLog2, sensorLog3, .. ]]
+ * where each element is a Realm.results<SensorLog>
+ */
+export const extractBreaches = ({ sensorLogs = [], database }) => {
+  if (!(database && sensorLogs.length > 0)) return [];
+
+  const sensorLogStack = [];
+  const breaches = [];
+
+  // Sort the sensor logs by timestamp.
+  sensorLogs.sorted('timestamp', true);
+
+  sensorLogs.forEach(sensorLog => {
+    const { isInBreach } = sensorLog;
+    const { length: stackLength } = sensorLogStack;
+    // If the stack is empty, just push this sensorLog.
+    if (stackLength === 0) sensorLogStack.push(sensorLog);
+    // Any breached sensorLogs are always pushed onto the stack.
+    else if (isInBreach) sensorLogStack.push(sensorLog);
+    // If both head of the stack and this sensorLog are unbreached,
+    // replace the head. This can occurs when the length is 1. Two
+    // sequential unbreached logs are never pushed.
+    else if (!sensorLogStack[stackLength - 1].isInBreach && !isInBreach) {
+      sensorLogStack.pop();
+      sensorLogStack.push(sensorLog);
+      // In all other cases, a breach has been found. Push the last sensorlog,
+      // store the breach and reset the stack. Push this sensorLog back
+      // onto the stack as the new potential delimiter for the next breach.
+    } else {
+      sensorLogStack.push(sensorLog);
+      breaches.push([...sensorLogStack]);
+      sensorLogStack.length = 0;
+      sensorLogStack.push(sensorLog);
+    }
+  });
+  // Push any remaining sensorlogs. If any are left, they form a breach
+  // with no delimiting non-breached sensorlog.
+  if (sensorLogStack.length > 0) breaches.push(sensorLogStack);
+
+  // Create Realm.result objects for each breach.
+  return breaches.map(breach => {
+    const sensorIds = breach.map(log => log.id);
+    const queryString = sensorIds.map((_, i) => `id = $${i}`).join(' OR ');
```

Comment:
why not breach.map({id} => `id = "${id}"`).join(' OR ')

Reply:
I tried this at first. It didn't work, can't remember the error exactly. Was something about having static predicates

Reply:
wanted to make sure: 
<img width="971" alt="image" src="https://user-images.githubusercontent.com/35858975/56359781-b0793400-6236-11e9-96a7-8f9ca378377a.png">

From what I can tell - queries were changed, if you had done this in the past. There are a few very old issues on the realm repo where this is suggested,  but doesn't work anymore.

Reply:
@joshxg, you are missing "

<img width="467" alt="Screen Shot 2019-04-19 at 8 15 48 PM" src="https://user-images.githubusercontent.com/26194949/56414844-f9d18e00-62df-11e9-8979-53a7807f0a9a.png">
<img width="1022" alt="Screen Shot 2019-04-19 at 8 15 14 PM" src="https://user-images.githubusercontent.com/26194949/56414842-f938f780-62df-11e9-882c-ce40ef795c87.png">

Reply:
if itemBatches.length is zero though, it crashes (empty query)

Reply:
.. 063e98340484f70afdc20b473b1edcab2936fdb0

- Won't ever have an empty query

Reply:
sure, sorry i wasn't clear enough, was just advising that the only crash with query would be if itemBatches.length is zero (which won't be the case)